### PR TITLE
Display error type (error, warning, script error) in OS::print_error + cleanup error ANSI coloring

### LINKED ISSUE #1127: Does not create editor_settings.xml when editor starts
Running godot_x11-1.0stable.32 for first time crates ~/.godot/plugins, ~/.godot/templates and ~/.godot/tmp are created during startup but does not create ~/.godot/editor_settings.xml then complains with error messages:
>
ERROR: create: Config file does not exist, creating.
  At: tools/editor/editor settings.cpp:234.
Error: scan_plugins: Plugin dir does not exist!
  At: tools/editor/editor_settings.cpp:319.

Observed with godot_x11-1.0stable.32 on Xubuntu 14.04 on Celeron M and Radeon Xpress 200M.
Also observed with godot_x11-1.0stable.64 on Xubuntu 14.10 on core i5.

### PR DESCRIPTION
**Pull request by akien-mga**
**tl;dr:** In the end, it looks like this:
- _Dark background colour:_
  gt-colour1
- _Light background colour:_
  gt-colour2

---

Previously all types of errors would be shown as ERROR, thus making for example warnings (WARN_PRINT) somewhat aggressive.

Fixes #1127.

Note that I'm a beginner C++ developer, so the coding logic I adopted might not be the best one. I also only tested the changes on Unix/X11, so the changes for Windows and WinRT need to be thoroughly reviewed (though they're more or less copy-pastes of the Unix code).

### CODE REVIEW

**Comment by akien-mga** [comment]
I also chose the colours at random:
* ERROR is in red, as previously
* WARNING is in yellow
* SCRIPT ERROR is in green

File: core/os/os.cpp [hunk 1]
```diff
@@ -61,9 +61,16 @@ void OS::debug_break() {
 
 void OS::print_error(const char* p_function,const char* p_file,int p_line,const char *p_code,const char*p_rationale,ErrorType p_type) {
 
+	const char* err_type;
+	switch(p_type) {
+		case ERR_ERROR: err_type="**WARNING**"; break;
```
Comment (volzhs) [inline core/os/os.cpp]: should this ``err_type`` be ``"**ERROR**"``?

Comment (akien-mga) [inline core/os/os.cpp]: Indeed, nice catch. I'll fix that.

Comment (akien-mga) [inline core/os/os.cpp]: This should now be fixed in bb6c3f5.

**Comment by akien-mga** [comment]
I added another commit to reset the terminal colour to the user setting instead of enforcing a white font after an error message.

**Comment by akien-mga** [comment]
Rebased the above commit to also remove the hardcoded background colour and use the terminal colour (in bold) instead of a hardcoded white colour for the error log.

In the end, it looks like this:
* Dark background colour:
gt-colour1
* Light background colour:
gt-colour2

**Comment by bojidar-bg** [comment]
:+1: Looks nice :smile:

**Comment by ZuBsPaCe** [comment]
Looks good on windows! (compiled with vs2015)

godot-error-types-windows

Some things I noticed:
* I'm not able to generate SCRIPT ERRORS without calling _err_print_error explicitly with that type. How do you generate those? As you can see in the screenshot, at the end I'm calling a method on a null-instance. Shouldn't this be a script error?
* In the screenshot I called print and printerr. In my opinion those should be colored, too. Maybe green and red?
* Green stands for 'ok', which does not work well with SCRIPT_ERROR... What about using purple or maybe orange?

**Comment by akien-mga** [comment]
> I'm not able to generate SCRIPT ERRORS without calling _err_print_error explicitly with that type. How do you generate those? As you can see in the screenshot, at the end I'm calling a method on a null-instance. Shouldn't this be a script error?

Yeah I added a formatting for script errors as they are part of the ErrorType enum, but I'm not sure they are really used right now. I would indeed expect your error in _ready() to be a script error, if @reduz can confirm maybe you could open a new issue to enable the use of script errors throughout the code?

> In the screenshot I called print and printerr. In my opinion those should be colored, too. Maybe green and red?

I don't think print should be coloured, but maybe printerr could be in red yes.

> Green stands for 'ok', which does not work well with SCRIPT_ERROR... What about using purple or maybe orange?

Ah yes that's a good point. I'll try purple or orange then.

**Comment by akien-mga** [comment]
@ZuBsPaCe The available colours are actually pretty limited: https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
I played around with some values, see e.g. here cyan and magenta:
gt-colour3

It's not perfect, but I guess I'll use magenta for the script error. Blue is also cool but the dark blue is way too dark to be readable on a black background.

**Comment by akien-mga** [comment]
I rebase to have a cleaner PR, the first commit does most of what was described initially plus uses magenta for script error instead of green.
The second commit adds some cosmetic improvements by removing some hardcoded font and background colours.

**Comment by vnen** [comment]
Looks good on Windows, yes, unless you set custom colors. The errors always use black background. 

windows-godot-console-color

I believe the Windows color coding is happening afterwards, since the prompt does not support ANSI codes. Check [these lines](https://github.com/godotengine/godot/blob/61c4c5795ff9057b1c67142758418a4f982a0241/platform/windows/os_windows.cpp#L1781-L1832).

**Comment by akien-mga** [comment]
Thanks for the test @vnen. I don't have experience with Windows, could you maybe see if you can find the parameters to make the output similar to that of ANSI-compatible terminals like in the screenshots in https://github.com/godotengine/godot/pull/2737#issuecomment-155347757?

**Comment by akien-mga** [comment]
Note that I did not change the colouring scheme for non ANSI-compatible Windows terminals though, so this is not a regression.

**Comment by ZuBsPaCe** [comment]
@akien-mga I like purple for script errors.

About the custom colors mentioned by @vnen. It's possible to determine the background color which can be used in the subsequent SetConsoleTextAttribute calls. So after basecol is set you do:

    uint32_t backcol = sbi.wAttributes & ~(FOREGROUND_RED|FOREGROUND_GREEN|FOREGROUND_BLUE|FOREGROUND_INTENSITY);

and then you need to add backcol to all SetConsoleTextAttribute calls except the last one, which resets the attributes, for example:

    SetConsoleTextAttribute(hCon,basecol|FOREGROUND_INTENSITY|backcol); 

Here's the result:

godot-error-types-windows-white

**Comment by vnen** [comment]
I was able to fix this. The following diff upon your current commits is working fine for me. Also remove escape sequences from windows, since it doesn't work and `\E` gives compilation warnings as non-recognized escape sequence (and put a missing `\n` in script error):

```diff
diff --git a/platform/windows/os_windows.cpp b/platform/windows/os_windows.cpp
index 964b9e7..579ddd0 100644
--- a/platform/windows/os_windows.cpp
+++ b/platform/windows/os_windows.cpp
@@ -1777,16 +1777,16 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 
 		switch(p_type) {
 			case ERR_ERROR:
-				print("\E[1;31mERROR: %s: \E[0m\E[1m%s\n",p_function,err_details);
-				print("\E[0;31m   At: %s:%i.\E[0m\n",p_file,p_line);
+				print("ERROR: %s: %s\n",p_function,err_details);
+				print("   At: %s:%i.\n",p_file,p_line);
 				break;
 			case ERR_WARNING:
-				print("\E[1;33mWARNING: %s: \E[0m\E[1m%s\n",p_function,err_details);
-				print("\E[0;33m     At: %s:%i.\E[0m\n",p_file,p_line);
+				print("WARNING: %s: %s\n",p_function,err_details);
+				print("     At: %s:%i.\n",p_file,p_line);
 				break;
 			case ERR_SCRIPT:
-				print("\E[1;35mSCRIPT ERROR: %s: \E[0m\E[1m",p_function,err_details);
-				print("\E[0;35m          At: %s:%i.\E[0m\n",p_file,p_line);
+				print("SCRIPT ERROR: %s: %s\n",p_function,err_details);
+				print("          At: %s:%i.\n",p_file,p_line);
 				break;
 		}
 	} else {
@@ -1794,17 +1794,18 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 		CONSOLE_SCREEN_BUFFER_INFO sbi; //original
 		GetConsoleScreenBufferInfo(hCon,&sbi);
 
-		SetConsoleTextAttribute(hCon,sbi.wAttributes);
-
-
+		WORD current_bg = sbi.wAttributes & (BACKGROUND_RED|BACKGROUND_GREEN|BACKGROUND_BLUE|BACKGROUND_INTENSITY);
+		WORD current_fg = sbi.wAttributes & (FOREGROUND_RED|FOREGROUND_GREEN|FOREGROUND_BLUE|FOREGROUND_INTENSITY);
 
 		uint32_t basecol=0;
 		switch(p_type) {
 			case ERR_ERROR: basecol = FOREGROUND_RED; break;
 			case ERR_WARNING: basecol = FOREGROUND_RED|FOREGROUND_GREEN; break;
-			case ERR_SCRIPT: basecol = FOREGROUND_GREEN; break;
+			case ERR_SCRIPT: basecol = FOREGROUND_RED|FOREGROUND_BLUE; break;
 		}
 
+		basecol |= current_bg;
+
 		if (p_rationale && p_rationale[0]) {
 
 			SetConsoleTextAttribute(hCon,basecol|FOREGROUND_INTENSITY);
@@ -1816,11 +1817,11 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 				case ERR_SCRIPT: print("SCRIPT ERROR: "); break;
 			}
 
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN|FOREGROUND_INTENSITY);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s\n",p_rationale);
 			SetConsoleTextAttribute(hCon,basecol);
 			print("At: ");
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s:%i\n",p_file,p_line);
 
 
@@ -1831,11 +1832,11 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 				case ERR_WARNING: print("WARNING: %s: ",p_function); break;
 				case ERR_SCRIPT: print("SCRIPT ERROR: %s: ",p_function); break;
 			}
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN|FOREGROUND_INTENSITY);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s\n",p_code);
 			SetConsoleTextAttribute(hCon,basecol);
 			print("At: ");
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s:%i\n",p_file,p_line);
 		}
```

As per WinRT, I don't think it has any console. Might be good to remove escape sequences there too, since it'll likely complain about `\E`.

Finally, might be needing some spaces before "At:" to be like the *nix version.
